components/layout_2020/display_list/stacking_context.rs

```diff
@@ -244,7 +244,107 @@ impl StackingContext {
+            // because the root element generates a stacking context. Let’s find that one.
+            self.stacking_contexts
+                .first()
+                .and_then(|first_child_stacking_context| {
+                    first_child_stacking_context.fragments.first()
+                })
+        });
+
+        let first_stacking_context_fragment = if let Some(first) = first_if_any {
+            first
+        } else {
+            // This should only happen if the root element has `display: none`
+            debug_panic!("`CanvasBackground::for_root_element` should have returned `style: None`");
+            return;
+        };
+
+        let fragment = first_stacking_context_fragment.fragment.borrow();
+        let box_fragment = if let Fragment::Box(box_fragment) = &*fragment {
+            box_fragment
+        } else {
+            debug_panic!("Expected a box-generated fragment");
+            return;
+        };
+
+        // The `StackingContextFragment` we found is for the root DOM element:
+        debug_assert_eq!(
+            box_fragment.tag,
+            fragment_tree.canvas_background.root_element
+        );
+
+        // Take the `SpatialId` to apply any CSS transform, but not the `ClipId`
+        builder.current_space_and_clip.spatial_id =
+            first_stacking_context_fragment.space_and_clip.spatial_id;
+
+        // The painting area is theoretically the infinite 2D plane,
+        // but we need a rectangle with finite coordinates.
+        //
+        // FIXME: a more accurate answer is:
+        //
+        // * Take two rectangles
+        //   - `fragment_tree.initial_containing_block` (area the viewport shows
+        //     when at initial scroll position), and
+        //   - `fragment_tree.scrollable_overflow` (areas we can scroll to)
+        // * Transform them from the top-level coordinate system
+        //   to the one for the local `SpatialId`.
+        //   This involves (see `build_reference_frame_if_necessary`)
+        //   - The change of origin to the top-left corner of the border rect of the fragment
+        //   - The result of `calculate_transform_matrix`
+        //   - The result of `calculate_perspective_matrix`
+        // * Take the axis-aligned
+        //   (in local coordinates, which may be rotated from top-level coordinates)
+        //   bounding box of those two transformed rectangles.
+        //
+        // Instead, let’s use a huge rectangle that will hopefully contain everything.
+        //
+        // `f32` has 24 bits of mantissa.
+        // 20 bits of magnitude leaves a few bits of precision
+        // for `background-position` computation to be accurate
+        // to more than a device pixel on very high DPI.
+        const BIG: f32 = (1u32 << 20) as f32;
+        let painting_area = &LayoutRect::new(
+            LayoutPoint::new(-BIG / 2., -BIG / 2.),
+            LayoutSize::new(BIG, BIG),
+        );
```

Comment:
```
  ▶ CRASH [expected FAIL] /_mozilla/css/mix_blend_mode_a.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  │ libEGL warning: No hardware driver found, falling back to software rendering
  │ [2020-05-07T12:17:09Z ERROR webrender::render_task] Attempting to create a render task of size 1048576x1048576
  │ explicit panic (thread WRRenderBackend#0, at /root/.cargo/git/checkouts/webrender-c3596abe1cf4f320/4ba17a6/webrender/src/render_task.rs:37)
```

Ok so using a huge rectangle might not be a good idea if WebRender wants to actually allocate that many pixels.

Reply:
Minimized reproduction:

```html
<html style="background: white">
<body style="mix-blend-mode: difference">
```

Reply:
The same panic happens for different inputs on `master`. I’ve filed https://github.com/servo/servo/issues/26463.